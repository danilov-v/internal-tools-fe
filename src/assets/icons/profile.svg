<svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="0 0 10.583 10.583"><defs><clipPath clipPathUnits="userSpaceOnUse" id="a"><rect width="11.67" height="4.654" x="174.18" y="10.52" ry="2.327" fill="#ccc" stroke="#ccc" stroke-width=".2" paint-order="fill markers stroke"/></clipPath></defs><g transform="matrix(2.79095 0 0 2.79095 -492.884 -17.371)" fill="#ccc" paint-order="fill markers stroke"><circle transform="rotate(180 132.942 7.537) scale(.48553)" clip-path="url(#a)" r="3.373" cy="10.589" cx="179.983"/><circle r=".782" cy="7.055" cx="178.497"/></g></svg>