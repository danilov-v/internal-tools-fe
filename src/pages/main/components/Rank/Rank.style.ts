import styled from 'styled-components';

export const RankWrapper = styled.div`
  position: relative;
  display: flex;
`;

export const RankImage = styled.img`
  width: 35px;
  height: 35px;
`;

export const RankCaption = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  color: #cccccc;
  margin-left: 5px;

  span {
    font-size: 10px;
    color: #727374;
  }
`;
