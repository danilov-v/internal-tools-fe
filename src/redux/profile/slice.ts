import { createSlice, SerializedError } from '@reduxjs/toolkit';

import { User } from 'types/user';
import { requestLogin, requestProfile, requestLogout } from './thunks';

export type ProfileState = {
  profile: User | null;
  error: SerializedError | null;
  isChecked: boolean;
};

const initialState: ProfileState = {
  profile: null,
  error: null,
  isChecked: false,
};

const profileSlice = createSlice({
  name: 'profile',
  initialState,
  reducers: {},
  extraReducers: builder => {
    builder.addCase(requestLogin.pending, state => {
      return {
        ...state,
        error: null,
      };
    });

    builder.addCase(requestLogin.fulfilled, (state, action) => {
      return {
        ...state,
        isChecked: true,
        profile: { ...action.payload },
      };
    });

    builder.addCase(requestLogin.rejected, (state, action) => {
      return {
        ...state,
        isChecked: true,
        profile: null,
        error: action.error,
      };
    });

    builder.addCase(requestProfile.fulfilled, (state, action) => {
      return {
        ...state,
        isChecked: true,
        profile: { ...action.payload },
        error: null,
      };
    });

    builder.addCase(requestProfile.rejected, (state, action) => {
      return {
        ...state,
        isChecked: true,
        profile: null,
        error: action.error,
      };
    });

    builder.addCase(requestLogout.fulfilled, (state, action) => {
      return {
        isChecked: true,
        profile: null,
        error: null,
      };
    });
  },
});

export { profileSlice };
