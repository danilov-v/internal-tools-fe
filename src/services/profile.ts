import { PROFILE_URL } from 'helpers/url';
import { fetchAPI } from 'services/fetch';
import { User } from 'types/user';

const getProfile = async (): Promise<User> => {
  const response = await fetchAPI<User>(PROFILE_URL);

  return response;
};

export { getProfile };
