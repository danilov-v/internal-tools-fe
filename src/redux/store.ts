import { configureStore, EnhancedStore } from '@reduxjs/toolkit';

import { appSlice, AppState } from './app/slice';
import { penaltySlice, PenaltyState } from './penalty/slice';
import { penaltyTypeSlice, PenaltyTypeState } from './penaltyType/slice';
import { personnelSlice, PersonnelState } from './personnel/slice';
import {
  personnelDetailsSlice,
  PersonnelDetailsState,
} from './personnel-details/slice';
import { positionSlice, PositionState } from './position/slice';
import { profileSlice, ProfileState } from './profile/slice';
import { promotionSlice, PromotionState } from './promotion/slice';
import { promotionTypeSlice, PromotionTypeState } from './promotionType/slice';
import {
  personnelRemovalTypeSlice,
  PersonnelRemovalTypeState,
} from './personnelRemovalType/slice';

import { rankSlice, RankState } from './rank/slice';
import { unitSlice, UnitState } from './unit/slice';
import { maritalStatusSlice, MaritalStatusState } from './maritalStatus/slice';

export const getStore = (initialState?: OptionalRootStore): EnhancedStore =>
  configureStore({
    reducer: {
      app: appSlice.reducer,
      penalty: penaltySlice.reducer,
      penaltyType: penaltyTypeSlice.reducer,
      personnel: personnelSlice.reducer,
      personnelDetails: personnelDetailsSlice.reducer,
      personnelRemovalType: personnelRemovalTypeSlice.reducer,
      position: positionSlice.reducer,
      profile: profileSlice.reducer,
      promotion: promotionSlice.reducer,
      promotionType: promotionTypeSlice.reducer,
      rank: rankSlice.reducer,
      unit: unitSlice.reducer,
      maritalStatus: maritalStatusSlice.reducer,
    },
    preloadedState: initialState,
  });

export type RootStore = {
  app: AppState;
  penalty: PenaltyState;
  penaltyType: PenaltyTypeState;
  personnel: PersonnelState;
  personnelDetails: PersonnelDetailsState;
  personnelRemovalType: PersonnelRemovalTypeState;
  position: PositionState;
  profile: ProfileState;
  promotion: PromotionState;
  promotionType: PromotionTypeState;
  rank: RankState;
  unit: UnitState;
  maritalStatus: MaritalStatusState;
};

export type OptionalRootStore = Partial<RootStore>;
