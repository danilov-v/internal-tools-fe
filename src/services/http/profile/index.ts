import { PROFILE } from 'configs/urls';
import { http } from 'services/http';
import { User } from 'types/user';

export const fetchProfile = async (): Promise<User> => http.get(PROFILE);
