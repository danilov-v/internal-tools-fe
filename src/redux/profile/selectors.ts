import { createSelector } from '@reduxjs/toolkit';
import { RootStore } from 'redux/store';
import { ProfileState } from './slice';

export const getProfileSlice = (state: RootStore): ProfileState =>
  state.profile;

export const getProfileInfo = createSelector(
  getProfileSlice,
  profileSliceData => profileSliceData.profile,
);

export const getProfileError = createSelector(
  getProfileSlice,
  profileSliceData => profileSliceData.error,
);

export const isAuthChecked = createSelector(
  getProfileSlice,
  profileSliceData => profileSliceData.isChecked,
);

export const isProfileExist = createSelector(
  getProfileSlice,
  profileSlice => profileSlice.profile,
);

export const isAuthorizedProfile = createSelector(
  isAuthChecked,
  isProfileExist,
  (isChecked, isExist) => isChecked && isExist,
);
