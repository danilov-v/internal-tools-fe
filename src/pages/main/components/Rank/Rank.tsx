import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';

import { getProfileInfo } from 'redux/profile/selectors';

import * as R from 'helpers/rank';

import captain from 'assets/ranks/captain.svg';
import lieutenant from 'assets/ranks/lieutenant.svg';
import seniorLieutenant from 'assets/ranks/senior-lieutenant.svg';

import * as S from './Rank.style';

const getRankImage = (rank: string): ReactElement => {
  switch (rank) {
    case R.CAPTAIN:
      return <S.RankImage src={captain} alt="Rank" />;
    case R.SENIOR_LIEUTENANT:
      return <S.RankImage src={seniorLieutenant} alt="Rank" />;
    case R.LIEUTENANT:
      return <S.RankImage src={lieutenant} alt="Rank" />;
    default:
      return <S.RankImage src={lieutenant} alt="Rank" />;
  }
};

export const Rank: React.FC = () => {
  const profile = useSelector(getProfileInfo);

  if (!profile) {
    return null;
  }

  return (
    <S.RankWrapper>
      {getRankImage('')}
      <S.RankCaption>
        {/* TODO: изменить profile.role на profile.position или что-то такое  */}
        <span>{profile.role}</span>
        {profile.lastName}
      </S.RankCaption>
    </S.RankWrapper>
  );
};
