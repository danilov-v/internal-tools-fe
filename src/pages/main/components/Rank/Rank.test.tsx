import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';

import { getStore } from 'redux/store';

import { EMPTY_PROFILE, FULFILLED_PROFILE } from 'tests/slices/profile';

import { Rank } from './Rank';

describe('<Rank />', () => {
  const getComponent = (initialStore = EMPTY_PROFILE): RenderResult => {
    const store = getStore(initialStore);

    return render(
      <Provider store={store}>
        <Rank />
      </Provider>,
    );
  };

  it('should render null when profile empty', () => {
    const component = getComponent();

    expect(component.baseElement).toMatchSnapshot();
  });

  it('should render and match snapshot profile exist', () => {
    const component = getComponent(FULFILLED_PROFILE);

    expect(component.baseElement).toMatchSnapshot();
  });
});
