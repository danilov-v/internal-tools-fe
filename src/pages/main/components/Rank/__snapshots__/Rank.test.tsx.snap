// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rank /> should render and match snapshot profile exist 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm eJEiKT"
    >
      <img
        alt="Rank"
        class="sc-AxirZ cDVCzV"
        src="lieutenant.svg"
      />
      <div
        class="sc-AxiKw eEODNF"
      >
        <span>
          admin
        </span>
        Gab
      </div>
    </div>
  </div>
</body>
`;

exports[`<Rank /> should render null when profile empty 1`] = `
<body>
  <div />
</body>
`;
